### PR TITLE
6676 port generative engines

### PR DESCRIPTION
**Pull request by marksgraham**
Part of #6676  .

### Description

A few sentences describing the changes proposed in this pull request.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/engines/trainer.py [hunk 1]
```diff
@@ -423,3 +423,282 @@ def _iteration(
             GanKeys.GLOSS: g_loss.item(),
             GanKeys.DLOSS: d_total_loss.item(),
         }
+
+
+class AdversarialTrainer(Trainer):
```
Comment (KumoLiu) [inline monai/engines/trainer.py:428]: I took a quick look and thought this class could be merged with `GanTrainer`. Alternatively, refactor `GanTrainer` into a more general base class. It appears that they all need to train a generator and discriminator.
What do you think?
cc @ericspod @Nic-Ma

Comment (ericspod) [reply monai/engines/trainer.py:428]: It would be a good idea to refactor GanTrainer, I don't think it's used much so there's room for improvement.

Comment (ericspod) [reply monai/engines/trainer.py:428]: Thinking about how we'd do adversarial training in general, I wonder if some approach where we have separate engines for each network, with training for the discriminator triggered by a handler when an iteration for the generator completes, and the double forward pass of the discriminator is done using a specialised inferer. It's possible we could eliminate specialised trainers for adversarial training completely, or at least make something more general that can handle things like CycleGAN or other more exotic setups.

Comment (marksgraham) [reply monai/engines/trainer.py:428]: Makes sense to refactor - I've added to the refactor list. 

I seem to be having a codeformat error, I think caused by an upgrade to black?

Comment (KumoLiu) [reply monai/engines/trainer.py:428]: Yes, have created a PR to fix it.
https://github.com/Project-MONAI/MONAI/pull/7422

Comment (marksgraham) [reply monai/engines/trainer.py:476]: ok merged in. not sure how to deal with DCO errors for other users, caused by merging in dev?

Comment (KumoLiu) [reply monai/engines/trainer.py:476]: Can the method "Rebase the branch" list here solve the issue?
https://github.com/Project-MONAI/MONAI/pull/7406/checks?check_run_id=21013930168

**Comment by marksgraham** [comment]
i think this is ready to go unless there are any other comments to address?

**Comment by KumoLiu** [comment]
> i think this is ready to go unless there are any other comments to address?

Yes, I will merge this one after [this one](https://github.com/Project-MONAI/MONAI/pull/7428).
@ericspod, Do you have any other comments here?

**Review by ericspod — APPROVED** [review]
I'm fine with this since the large number of modified files is mostly integrated changes. We'll be doing other code reviews to catch anything here later.

**Pull request merged**